packages/components/src/autocomplete/types.ts

```diff
@@ -0,0 +1,214 @@
+ */
+import type { useAutocomplete } from '.';
+
+type GetOptions< TCompleterOption = any > = (
+	filterValue: AutocompleterUIProps[ 'filterValue' ]
+) => Array< TCompleterOption > | Promise< Array< TCompleterOption > >;
+
+export type OptionCompletion = {
+	action: 'insert-at-caret' | 'replace';
+	value: string | WPElement;
+};
+
+type getOptionCompletion< TCompleterOption = any > = (
+	option: TCompleterOption,
+	query: string
+) => OptionCompletion | OptionCompletion[ 'value' ];
+
+type OptionLabel = string | WPElement | Array< string | WPElement >;
+export type KeyedOption = {
+	key: string;
+	value: any;
+	label: OptionLabel;
+	keywords: Array< string >;
+	isDisabled: boolean;
+};
+
+export type WPCompleter< TCompleterOption = any > = {
+	/**
+	 * The name of the completer. Useful for identifying a specific completer to
+	 * be overridden via extensibility hooks.
+	 */
+	name: string;
+	/**
+	 * The string prefix that should trigger the completer. For example,
+	 * Gutenberg's block completer is triggered when the '/' character is entered.
+	 */
+	triggerPrefix: string;
+	/**
+	 * The raw options for completion. May be an array, a function that returns
+	 * an array, or a function that returns a promise for an array.
+	 * Options may be of any type or shape. The completer declares how those
+	 * options are rendered and what their completions should be when selected.
+	 */
+	options: Array< TCompleterOption > | GetOptions;
+	/**
+	 * A function that returns the keywords for the specified option.
+	 */
+	getOptionKeywords?: ( option: TCompleterOption ) => Array< string >;
+	/**
+	 * A function that returns whether or not the specified option is disabled.
+	 * Disabled options cannot be selected.
+	 */
+	isOptionDisabled?: ( option: TCompleterOption ) => boolean;
+	/**
+	 * A function that returns the label for a given option. A label may be a
+	 * string or a mixed array of strings, elements, and components.
+	 */
+	getOptionLabel: ( option: TCompleterOption ) => OptionLabel;
+	/**
+	 * A function that takes a Range before and a Range after the autocomplete
+	 * trigger and query text and returns a boolean indicating whether the
+	 * completer should be considered for that context.
+	 */
+	allowContext?: ( before: string, after: string ) => boolean;
+	/**
+	 * A function that takes an option and returns how the option should
+	 * be completed. By default, the result is a value to be inserted in the text.
+	 * However, a completer may explicitly declare how a completion should be
+	 * treated by returning an object with `action` and `value` properties. The
+	 * `action` declares what should be done with the `value`.
+	 */
+	getOptionCompletion?: getOptionCompletion;
+	/**
+	 * A function that returns an array of items to be displayed in the
+	 * Autocomplete UI. These items have uniform shape and have been filtered by
+	 * `AutocompleterUIProps.filterValue`.
+	 */
+	useItems?: ( filterValue: string ) => [ Array< KeyedOption > ];
```

Comment:
It is intentional... this one was a bit of a rabbit hole for me actually. I probably should have left a note on it.

`filteredOptions` [returns an array](https://github.com/WordPress/gutenberg/blob/057c37c5ff20b8e5c8b4552143f0ff45a5b41712/packages/components/src/autocomplete/get-default-use-items.tsx#L18) of `KeyedOptions`.

`getDefaultUseItems` uses `filteredOptions` [to update its list of items](https://github.com/WordPress/gutenberg/blob/057c37c5ff20b8e5c8b4552143f0ff45a5b41712/packages/components/src/autocomplete/get-default-use-items.tsx#L104) in a layout effect. From there, `getDefaultUseItems` [wraps that array in an outer array as its return value](https://github.com/WordPress/gutenberg/blob/057c37c5ff20b8e5c8b4552143f0ff45a5b41712/packages/components/src/autocomplete/get-default-use-items.tsx#L122).

The really fun part is that when `getAutoCompleterUI` ultimately calls `getDefaultUseItems` (here being called under a different variable name), [it destructures the original array back out again](https://github.com/WordPress/gutenberg/blob/a3f56ce01fe7deddbd5153decdebf0b9dc656c3f/packages/components/src/autocomplete/autocompleter-ui.tsx#L40).

So basically, an array is cast as a tuple to serve as a return value, and then immediately destructured back to an array when the function executes. I'm not sure why array is tupled then immediately de-tupled (I don't think that's a word but I'm rolling with it 😆 ) but it's only ever one array of `KeyedOptions` and even then it never exists long enough to be acted upon.

What all of this boils down to (unless I'm mistaken, which is always possible) is that if a consumer _did_ end up providing a `useItems` function prop that returned more than one `Array< KeyedOption>` the first one would get destructured to the `items` variable, and everything else would be ignored.

Reply:
Expanding on this one: it would be interesting to git blame and research the original reasons for why this is expected to be a single element tuple, and see if it's possible to remove what feels like an unnecessary impediment... but given the nature of a TS migration, it's definitely not part of this PR's scope.